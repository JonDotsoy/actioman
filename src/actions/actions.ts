import type { z } from "zod";
import type { Configs, ConfigsModule } from "../configs/configs.js";
import { get } from "@jondotsoy/utils-js/get";
import actionsList from "../share-actions/share-actions.js";

export const actions = actionsList;
const GeneratorFunction = function* () {}.constructor;
const AsyncGeneratorFunction = async function* () {}.constructor;

type InferZodType<T> = T extends z.ZodType ? z.infer<T> : any;
type ParameterTypeInferer<T> = T extends (i: infer R) => any ? R : any;
type Handler<
  InputType = any,
  ResultType = any,
  SSE extends boolean = any,
> = SSE extends true
  ? (
      input: InferZodType<InputType>,
    ) =>
      | Generator<InferZodType<ResultType>>
      | AsyncGenerator<InferZodType<ResultType>>
  : (
      input: InferZodType<InputType>,
    ) => InferZodType<ResultType> | Promise<InferZodType<ResultType>>;

export type Action<
  InputType = any,
  ResultType = any,
  SSE extends boolean = false,
  T = any,
> = {
  description?: string;
  input?: InputType;
  output?: ResultType;
  sse?: SSE;
  handler: Handler<InputType, ResultType, SSE>;
};

export const defineAction = <I = any, R = any, SSE extends boolean = false>(
  action: Action<I, R, SSE>,
): Action<I, R, SSE> => action;

export const actionFromModule = (module: any) => {
  const handlerFn = get.function(module) as
    | undefined
    | ((...args: any[]) => any);
  if (handlerFn) {
    const isGenerator =
      handlerFn instanceof GeneratorFunction ||
      handlerFn instanceof AsyncGeneratorFunction;
    return defineAction({ sse: isGenerator, handler: handlerFn });
  }
  const handlerObj = get.record(module);
  if (!handlerObj) return null;
  const handler = get.function(handlerObj, "handler") as
    | undefined
    | ((...args: any[]) => any);
  if (!handler) return null;
  return defineAction({
    description: get.string(handlerObj, "description"),
    input: get.record(handlerObj, "input"),
    output: get.record(handlerObj, "output"),
    handler,
  });
};

export class Actions<ActionsDefinitions extends Record<string, Action> = any> {
  constructor(private definition: ActionsDefinitions) {}

  /** @deprecated It has never been used */
  call = async <NameActionDefinition extends keyof ActionsDefinitions>(
    name: NameActionDefinition,
    input: ParameterTypeInferer<
      ActionsDefinitions[NameActionDefinition]["handler"]
    >,
  ): Promise<
    ReturnType<ActionsDefinitions[NameActionDefinition]["handler"]>
  > => {
    const definition = this.definition[name];
    if (!definition) throw new Error(`Action ${name.toString()} not found`);
    return await definition.handler(input);
  };

  static describe(actions: Actions): Record<string, Action> {
    return actions.definition;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const isRecord =
      typeof module === "object" && module !== null && !Array.isArray(module);

    if (!isRecord) return null;

    const actionDefinitions: Record<string, Action<any, any, any>> = {};

    for (const [moduleName, handler] of Object.entries(module)) {
      const definition = actionFromModule(handler);
      if (!definition) break;
      actionDefinitions[moduleName] = definition;
    }

    return new Actions(actionDefinitions);
  }
}
