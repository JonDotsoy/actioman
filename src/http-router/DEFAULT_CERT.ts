export const DEFAULT_CERT =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIIDCTCCAfGgAwIBAgIUcx4/uMJz+em1t9pyg8YBn8mOtz8wDQYJKoZIhvcNAQEL\n" +
  "BQAwFDESMBAGA1UEAwwJbG9jYWxob3N0MB4XDTI1MDIwNzE1MDcyMloXDTI1MDMw\n" +
  "OTE1MDcyMlowFDESMBAGA1UEAwwJbG9jYWxob3N0MIIBIjANBgkqhkiG9w0BAQEF\n" +
  "AAOCAQ8AMIIBCgKCAQEApcFu5mawUooMhsH/1bYdRMn1PTlmRiPRJkoookkWOhVH\n" +
  "703iYVCkZL39DF9tXG4iANNdjFepVIcSZduyZBXslAvuvFWaNu9I52IYyDYDadDT\n" +
  "oLvXFi3dsfnIrGf68JLBqZnOSxVLg80Ht3maOSbc36DbUO5YE9nU0nhA/wQ983j0\n" +
  "FDzBKzrPYPFNIbmI3SEVPjhrFrxci7+Tbdky4SpIwLRN7SWpbu/GPmLzlvD8XMuW\n" +
  "cw1dsBoqV9w+pMwy2DTctQ3SU3nbxuvMgMKu4wuX9mWLiCmykGcLvYUrI8CKFzzP\n" +
  "f48Qd2/HC1lzwCKiwn57MhokbzcKWhB8E96X3HGtVwIDAQABo1MwUTAdBgNVHQ4E\n" +
  "FgQUPsTW/85jbxCCvyiMwx33xM6nh/IwHwYDVR0jBBgwFoAUPsTW/85jbxCCvyiM\n" +
  "wx33xM6nh/IwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEANfYP\n" +
  "er5WrTFuWtY2ApxMjvDWo0xuL6+kCv4UkaLtfsR4TkcF9svGtb/PvrlOmffP+By1\n" +
  "TkBoMju3uHU7/hPo5UMRUI445QnIVSSpLaHnLHBi85KGMl55ErRbOGVgYR1n+Qxo\n" +
  "nFwucn2hjHBmswXNY9DJx8vmewKqa9Kc694YxNx44tmMNbokUYKliofnisxxhM3Y\n" +
  "4REA+w3bme7z7JDLt0J1EGCAEeT2NcLbnXw8gEKJ0D39Zt5b5qR1Re4wWtknXlkc\n" +
  "5VGw96c5jbejD49BJsDWqVsJG1D4dhJ61ybBT6j2aRnSH+MxPHpH/Rm9wXIGha4G\n" +
  "mZLuodbHDBrL/FT4cQ==\n" +
  "-----END CERTIFICATE-----\n";
