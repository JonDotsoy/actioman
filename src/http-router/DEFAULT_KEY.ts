if (process.env.NODE_ENV !== "test") {
  console.warn("The variable DEFAULT_KEY is only to test environment");
}

export const DEFAULT_KEY =
  "-----BEGIN PRIVATE KEY-----\n" +
  "MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQClwW7mZrBSigyG\n" +
  "wf/Vth1EyfU9OWZGI9EmSiiiSRY6FUfvTeJhUKRkvf0MX21cbiIA012MV6lUhxJl\n" +
  "27JkFeyUC+68VZo270jnYhjINgNp0NOgu9cWLd2x+cisZ/rwksGpmc5LFUuDzQe3\n" +
  "eZo5JtzfoNtQ7lgT2dTSeED/BD3zePQUPMErOs9g8U0huYjdIRU+OGsWvFyLv5Nt\n" +
  "2TLhKkjAtE3tJalu78Y+YvOW8Pxcy5ZzDV2wGipX3D6kzDLYNNy1DdJTedvG68yA\n" +
  "wq7jC5f2ZYuIKbKQZwu9hSsjwIoXPM9/jxB3b8cLWXPAIqLCfnsyGiRvNwpaEHwT\n" +
  "3pfcca1XAgMBAAECggEAGE8voAvEZT6IysSt7B5a2YlS9aky5it/k4GBxxc38snH\n" +
  "NIH0JviIO5wffg6Kv+kJBJz7FcePdm2e1aBsX8BVY3sChs15IgBi7UhWQRqm0oCF\n" +
  "YQfbag3M+zqaYyqWw2yKiYCcoCdd3VUXAiIL68DW2NdDAvIJqdOjIkDvq1bobQn+\n" +
  "MNdNBnu58t95JRfEbQHiDUUVTGv1b4E8gAUtCmsDPtsSQAox8rTD4UTA0+9oxE5e\n" +
  "SoV69YWS+BqFaDZgxxTKxyoL6g4K7WrF58oOSIlayUBGOO0VpGBf3sFB9YvYw76P\n" +
  "7m8S1EY71G8/Bp4JoVrNR7TemG0Qotc/3AapEjyv7QKBgQDVKfOWe81oKuFJlpKP\n" +
  "1W0v7xoo0fV9GXJlOrpDTvcuoY+XiMuQtTWDl7de+4y+mzQ+YUWKCrEn6yj0eUgy\n" +
  "owEpk8p6E+nvU0w9Yhnlk0wmfHkQw/P4kwCuG7sFvcpJ8471Ad94NWLddeHdVQZb\n" +
  "3A8PZgXouoBstTlKJ9WkZHxmCwKBgQDHEJmiR/U6m882A8G7yhoDB4U2tYPS2WZq\n" +
  "sZdgaCDrYZNME46N11b6HEJYD6jHyP33FPgyLD9uz157/1WTod6y2fdrlRnTWXZt\n" +
  "KKX837tOaKZFYpJtTzVFdR6nCO6O4fL6tK/CyNIWOB+kAXmtr8GBfbBfGHuePuYr\n" +
  "9jjkGmIhZQKBgBes7Cgnof3S3Wjd93NQQitTxDUKdNirLeB+cFevepbMTPWItQti\n" +
  "hQoTnMNI7y3wcRIInb+iRmDnponYuqMdJbDngFKkyZPSTaJ3teagwJF5u7yX9w4B\n" +
  "zc6It/IATWp/Sb5+scJZM5zjTe3WVrl3LJkKPNfHdyIb0rwgxoLWwlZXAoGALI8s\n" +
  "tMTCwgHKml6w6SZOOEtgqGRtQ69iW5OxPOVSKR5IaIkgKeQAFjBpMWkf8MivALXy\n" +
  "8Ngyws0sazk2oLtlLEfaOHzBsSufdMPalAZI2T5XqcRs3ocSd5/F4sH/FIDDcRct\n" +
  "GL73N/4MJxwlbpmfCStnZYIHed6+iyysg3LccHkCgYBTPy0YejvOaTr3FMqIeL4r\n" +
  "3TsSbN9Q3/BX0oEvgTiJvLU1wsT3G0VzBE9qDbAmn9NuJhDKKgQRA7BdPVijZwib\n" +
  "NS49DtF2a5rB2wXfWNTG4KsQUPQ+zkhu3TyFUQijdlD+Vh47vD3bnH8hgInllUNc\n" +
  "WWPRkIP/TR8122xeYnv/hQ==\n" +
  "-----END PRIVATE KEY-----\n";
