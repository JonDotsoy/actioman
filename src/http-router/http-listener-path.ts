export const HTTP_LISTENER_PATH = new URL(
  "./http-listener.js",
  import.meta.url,
);
