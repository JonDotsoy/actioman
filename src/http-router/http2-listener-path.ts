export const HTTP2_LISTENER_PATH = new URL(
  "./http2-listener.js",
  import.meta.url,
);
