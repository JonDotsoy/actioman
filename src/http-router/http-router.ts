import { Router } from "artur";
import { Actions, type Action } from "../actions/actions.js";
import { z } from "zod";
import { get } from "@jondotsoy/utils-js/get";
import { result } from "@jondotsoy/utils-js/result";
import { actionsToJson } from "../exporter-actions/exporter-actions.js";
import { Configs, type ConfigsModule } from "../configs/configs.js";
import { MetricsClient } from "../metric/metrics-client/metrics-client.js";

const isParser = (value: any): value is { parse: (value: any) => any } =>
  get.function(value, "parse") !== undefined;
const isZodType = (value: any): value is z.ZodType =>
  value instanceof z.ZodType;
const isSSEAction = (
  action: Action<any, any, any>,
): action is Action<any, any, true> => action.sse === true;

class EventStreamDataEncoder {
  encode(data: { id?: string; event?: string; data: any }) {
    const l = [];
    if (data.id) l.push(`id: ${data.id}`);
    if (data.event) l.push(`event: ${data.event}`);
    l.push(`data: ${JSON.stringify(data.data)}`);
    return new TextEncoder().encode(l.join("\n") + "\n\n");
  }
}

export class HTTPRouter {
  router: Router<"pass">;
  metrics: MetricsClient;

  constructor(
    actions: Actions,
    readonly configs?: Configs,
  ) {
    const actionsJson = actionsToJson(actions);

    this.metrics = new MetricsClient();

    this.router = new Router({
      errorHandling: "pass",
      middlewares: [
        ...(configs?.httpListenerMiddlewares ?? []),
        (fetch) => async (req) => {
          const res = await fetch(req);
          res?.headers.set("X-Powered-By", "Actioman");
          res?.headers.set("X-Actioman-Version", "v1.0.0");
          return res;
        },
      ],
    });

    configs?.httpSetup(this);

    this.router.use("GET", `/__actions`, {
      fetch: () => Response.json({ actions: actionsJson }),
    });

    for (const [name, describe] of Object.entries(Actions.describe(actions))) {
      const metricLabels = {
        action: name,
      };

      const requestErrorCountersMetric = this.metrics.counter({
        name: "request_error_counters",
        labels: metricLabels,
      });
      const requestCountersMetric = this.metrics.counter({
        name: "request_counters",
        labels: metricLabels,
      });
      const requestDataTransferenceBytesMetric = this.metrics.counter({
        name: "request_data_transference_bytes",
        labels: metricLabels,
      });
      const requestDurationSecondsMetric = this.metrics.averageBySecond({
        name: "request_duration_seconds",
        labels: metricLabels,
      });

      const isSSE = isSSEAction(describe);
      const method = "POST";
      this.router.use(method, `/__actions/${name}`, {
        middlewares: [
          (fetch) => (req) => {
            const start = Date.now();
            requestCountersMetric.increment();
            return Promise.resolve(fetch(req))
              .catch((err) => {
                requestErrorCountersMetric.increment();
                throw err;
              })
              .finally(() => {
                requestDurationSecondsMetric.add(Date.now() - start);
              });
          },
        ],
        fetch: async (req) => {
          if (isSSE) return this.sseHandler(describe, req);
          return this.handler(describe, req);
        },
      });
    }
  }

  private async sseHandler(action: Action<any, any, true>, req: Request) {
    const url = new URL(req.url);
    const inputSearchParams = url.searchParams.get("input") ?? null;
    const [, data] = await result(() =>
      inputSearchParams ? JSON.parse(inputSearchParams) : null,
    );
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultAction = await action.handler(input);
    const body = new ReadableStream<Uint8Array>({
      pull: async (ctrl) => {
        const [err, iteratorResult] = await result(() => resultAction.next());

        if (err) {
          console.error(err);
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "error",
              data: "Internal error",
            }),
          );
          return ctrl.close();
        }

        const { done, value } = iteratorResult;

        if (done) {
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "close",
              data: true,
            }),
          );
          return ctrl.close();
        }

        const output = isParser(action.output)
          ? action.output.parse(value)
          : value;

        ctrl.enqueue(new EventStreamDataEncoder().encode({ data: output }));
      },
    });
    return new Response(body, {
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
        "X-Accel-Buffering": "no",
      },
    });
  }

  private async handler(action: Action<any, any, any>, req: Request) {
    const [, data] = await result(() => req.json());
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultDescribe = await action.handler(input);
    const output = isParser(action.output)
      ? action.output.parse(resultDescribe)
      : resultDescribe;
    return Response.json(output);
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const actions = Actions.fromModule(module, configs);
    if (!actions) throw new Error(`No actions found in ${module}`);
    return new HTTPRouter(actions, Configs.fromModule(configs));
  }
}
