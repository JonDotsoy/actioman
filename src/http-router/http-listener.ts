import * as http from "http";
import { HTTPRouter } from "./http-router.js";
import net from "net";
import type { ConfigsModule } from "../configs/configs.js";

type ListenOptions = {
  silent: boolean;
};

const sanatizeHostname = (hostname: string) => {
  switch (hostname) {
    case "::1":
    case "::":
    case "0.0.0.0":
    case "127.0.0.1":
    case "::ffff:127.0.0.1":
      return "localhost";
  }

  return hostname;
};

const INITIAL_PORT = 30_320;

const findNextPort = async () => {
  let porposalPort = INITIAL_PORT;
  while (true) {
    porposalPort++;
    if (porposalPort >= INITIAL_PORT + 10_000)
      throw new Error("No available port");
    const port = await new Promise<number | null>((resolve) => {
      const connectiong = net.connect({
        host: "localhost",
        port: porposalPort,
      });

      connectiong.addListener("connect", () => {
        resolve(null);
        connectiong.destroy();
      });
      connectiong.addListener("error", (err) => {
        if ("code" in err && err.code === "ECONNREFUSED") resolve(porposalPort);
      });
    });
    if (typeof port === "number") return port;
  }
};

export class HTTPLister {
  server: http.Server;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http.createServer(async (req, res) => {
      try {
        if (!(await httpRouter.router.requestListener(req, res))) {
          res.statusCode = 404;
          return res.end();
        }
      } catch (error) {
        console.error(error);
        res.statusCode = 500;
        return res.end();
      }
    });
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close((err) => {
        if (err) return reject(err);
        return resolve(null);
      });
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTPLister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "localhost";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `http://${sanatizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `http://${sanatizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTPLister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}
