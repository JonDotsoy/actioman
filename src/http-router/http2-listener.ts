import { HTTPRouter } from "./http-router.js";
import type { ConfigsModule } from "../configs/configs.js";
import { sanitizeHostname } from "./utils/sanitize-hostname.js";
import { findNextPort } from "./utils/find-next-port.js";
import * as http2 from "http2";
import { DEFAULT_CERT } from "./DEFAULT_CERT.js";
import { DEFAULT_KEY } from "./DEFAULT_KEY.js";
import { requestHttp2ToRequest } from "./utils/request-http2-to-request.js";

type ListenOptions = {
  silent: boolean;
};

export const INITIAL_PORT = 30_320;

export class HTTP2Lister {
  server: http2.Http2Server;
  url: URL | undefined;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http2.createSecureServer(
      {
        key: DEFAULT_KEY,
        cert: DEFAULT_CERT,
        settings: { initialWindowSize: 8 * 1024 * 1024 },
      },
      async (requestHttp2, responseHttp2) => {
        try {
          // console.log(r)

          const response = await this.httpRouter.router.fetch(
            requestHttp2ToRequest(requestHttp2),
          );

          if (!response) {
            responseHttp2.statusCode = 404;
            return responseHttp2.end();
          }

          responseHttp2.statusCode = response.status;
          response.headers.forEach((value, key) => {
            if (key.startsWith(":")) return;
            if (responseHttp2.hasHeader(key)) {
              responseHttp2.appendHeader(key, value);
            } else {
              responseHttp2.setHeader(key, value);
            }
          });
          if (response.body) {
            const reader = response.body.getReader();
            for (
              let readResult = await reader.read();
              !readResult.done && !requestHttp2.aborted;
              readResult = await reader.read()
            ) {
              responseHttp2.write(readResult.value);
            }
          }
          return responseHttp2.end();
        } catch (error) {
          console.error(error);
          responseHttp2.statusCode = 500;
          return responseHttp2.end();
        }
      },
    );
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close();
      return resolve(null);
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTP2Lister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "::";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `https://${sanitizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `https://${sanitizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    this.url = url;

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTP2Lister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}

export const HTTPLister = HTTP2Lister;
