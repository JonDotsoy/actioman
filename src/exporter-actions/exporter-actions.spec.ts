import { describe, it, expect, afterEach } from "bun:test";
import { Actions, defineAction } from "../actions/actions";
import { z } from "zod";
import {
  ActionsDocument,
  actionsToJson,
  importActionsFromJson,
} from "./exporter-actions";
import type { ActionsDefinitionsJsonDTO } from "./dtos/actions-definitions-json.dto";
import { HTTPLister } from "../http-router/http-listener";

const cleanupTasks = new Set<() => any>();

describe("actionsToJson", () => {
  it("should export actions to JSON", () => {
    const actions = new Actions({
      hi: defineAction({
        description: "Say hello",
        input: z.object({
          name: z.string(),
        }),
        output: z.string(),
        handler: async ({ name }) => `hello ${name}`,
      }),
    });

    expect(actionsToJson(actions)).toMatchSnapshot();
  });
});

afterEach(async () => {
  for (const cleanupTask of cleanupTasks) await cleanupTask();
});

describe("importActionsFromJson", () => {
  afterEach(async () => {
    for (const cleanupTask of cleanupTasks) await cleanupTask();
  });

  it("should import actions from JSON", () => {
    expect(importActionsFromJson({}, "http://localhost")).toMatchSnapshot();
  });

  it("should import actions from JSON with different types", () => {
    expect(
      importActionsFromJson(
        {
          hi: {
            description: "Say hello",
            input: {
              type: "object",
              required: ["name"],
              properties: {
                name: {
                  type: "string",
                },
              },
            },
            output: {
              type: "string",
            },
          },
          foo: {
            description: null,
            input: null,
            output: null,
          },
        } satisfies ActionsDefinitionsJsonDTO,
        "http://localhost:9080",
      ),
    ).toMatchSnapshot();
  });
});

it("should create ActionsDocument from HTTP server", async () => {
  const httpLister = HTTPLister.fromModule({
    hi: {
      input: z.object({
        name: z.string().optional(),
        metadata: z.record(z.string()),
      }),
      handler: () => "ok",
    },
    foo: {
      description: "foo",
      input: z.string(),
      output: z.object({
        name: z.string(),
      }),
      handler: () => "ok",
    },
  });
  cleanupTasks.add(() => httpLister.close());
  const url = await httpLister.listen(43112);

  const actionsDocument = await ActionsDocument.fromHTTPServer(url);
  expect(actionsDocument.toString()).toMatchSnapshot();
});
