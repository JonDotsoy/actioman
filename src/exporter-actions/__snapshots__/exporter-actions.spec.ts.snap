// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`actionsToJson should export actions to JSON 1`] = `
{
  "hi": {
    "description": "Say hello",
    "input": {
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
    "output": {
      "type": "string",
    },
  },
}
`;

exports[`importActionsFromJson should import actions from JSON 1`] = `
"import { z } from "zod";

"
`;

exports[`importActionsFromJson should import actions from JSON with different types 1`] = `
"import { z } from "zod";

export const hi = {
  description: "Say hello",
  target: "http://localhost:9080/__actions/hi"
  input: z.object({ "name": z.string() }),
  output: z.string(),
} as const;

export const foo = {
  description: undefined,
  target: "http://localhost:9080/__actions/foo"
  input: null,
  output: null,
} as const;

"
`;

exports[`should create ActionsDocument from HTTP server 1`] = `
"// @ts-nocheck
import { z } from 'zod';
import { ActionsTarget } from './node_modules/actions-target/actions-target.js';
const createActionsTarget = () => new ActionsTarget('http://localhost:43112/__actions', {
    hi: {
        description: undefined,
        input: z.object({ "name": z.string().optional(), "metadata": z.record(z.string()) }).strict(),
        output: undefined
    },
    foo: {
        description: 'foo',
        input: z.string(),
        output: z.object({ "name": z.string() }).strict()
    }
}).compile();
export default createActionsTarget;"
`;
